onnx/defs/rnn/defs.cc

```diff
@@ -6,125 +6,206 @@
-           "For LSTM only: optional initial value of the cell. If not "
-           "specified - assumed to be 0. Dimensions `[num_layers * directions, "
-           "batch_size, hidden_size]`", "T")
-    .Output(0, "output", "The output 3-dim sequence.", "T")
-    .Output(1, "output_h",
-            "Optional output value of the hidden. Same shape as input_h", "T")
-    .Output(2, "output_c",
-            "For LSTM only: optional output value of the cell. Same shape as "
-            "input_h", "T")
-    .TypeConstraint("T", { "tensor(float16)", "tensor(float)", "tensor(double)" }, "Constrain input and output types to float tensors.");
+Equations (forward LSTM with default activations and peepholes):
+  - it = sigmoid(Wi*Xt + Ri*Ht-1 + Pi (.) Ct-1 + Wbi + Rbi)
+  - ft = sigmoid(Wf*Xt + Rf*Ht-1 + Pf (.) Ct-1 + Wbf + Rbf)
+  - ct = tanh(Wc*Xt + Rc*Ht-1 + Wbc + Rbc)
+  - Ct = ft (.) Ct-1 + it (.) ct
+  - ot = sigmoid(Wo*Xt + Ro*Ht-1 + Po (.) Ct + Wbo + Rbo)
+  - H = ot (.) tanh(Ct)
+)DOC")
+    .Attr("activations", "A list of 4 (or 8 if bidirectional) activation functions "
+          "for input, output, forget, and cell gates. The activation functions must "
+          "be one of sigmoid and tanh. See the equations for default.",
+          AttrType::STRINGS)
+    .Attr("clip", "Cell clip threshold. Clipping bounds the elements of a tensor "
+          "in the range of [-threshold, +threshold] and is applied to the input "
+          "of activations. No clip if not specified.",
+          AttrType::FLOAT)
+    .Attr("input_forget", "Couple the input and forget gates if 1, default 0.",
+          AttrType::INT)	  
+    .Input(1, "W",
+	   "The weight tensor for the gates. Concatenation of `W[iofc]` and "
+           "`WB[iofc]` (if bidirectional) along dimension 0. The tensor has shape "
+	   "`[num_directions, 4*hidden_size, input_size]`.", "T")
+    .Input(2, "R",
+	   "The recurrence weight tensor. Concatenation of `R[iofc]` and "
+	   "`RB[iofc]` (if bidirectional) along dimension 0. This tensor has shape "
+           "`[num_directions, 4*hidden_size, hidden_size]`.", "T")
+    .Input(3, "B",
+	   "The bias tensor for input gate. Concatenation of `[Wb[iofc], Rb[iofc]]`, "
+	   "and `[WBb[iofc], RBb[iofc]]` (if bidirectional) along dimension 0. This "
+           "tensor has shape `[num_directions, 8*hidden_size]`. Optional: If not "
+	   "specified - assumed to be 0.", "T",
+	   true /*optional*/)
+    .Input(6, "initial_c",
+           "Optional initial value of the cell. If not specified - assumed "
+	   "to be 0. It has shape `[num_directions, batch_size, hidden_size]`.",
+	   "T", true /*optional*/)
+    .Input(7, "P",
+	   "The weight tensor for peepholes. Concatenation of `P[iof]` and "
+	   "`PB[iof]` (if bidirectional) along dimension 0. It has shape "
+	   "`[num_directions, 3*hidde_size]`. Optional: If not specified - "
+	   "assumed to be 0.", "T",
+	   true /*optional*/)
+    .FillUsing(RNNDocGenerator("LSTM"));
```

Comment:
sorry for nitpicking, but the order of arguments is weird in this case:

X,W,R,B
initial_h
seq_lengths
initial_c
P

Maybe group them together? At least seq_len shouldn't separate out init_c and init_h. As for P it's fine as it's rarely used, so we can keep it last

Reply:
@yuanbyu - would you mind to address this comment? 'Conditional accept' assumes that issues are addressed before merging :)

Reply:
Sure. I will have a small PR for this coming your way. :-)